### PR TITLE
feat: [M3-8332] - Use new "lish" API instead of "lish_token"

### PR DESCRIPTION
**Pull request by plisiecki1**
## Description 📝
This change upgrades the backend weblish/glish session creation requests from the older `lish_token` API to a newer `lish` API. This newer API includes support for both core and distributed regions.

## Changes  🔄
List any change relevant to the reviewer.
- This change uses the new `lish` API which returns the full URLs rather than leaving it up to cloud manager to compute. For core regions, this should result in the same URL as cloud manager would have computed.
- The new `lish` API also returns values to use as WebSocket "protocols"; this will currently be an empty list for core regions, but include session information for distributed regions.

## Target release date 🗓️
ASAP

## Preview 📷
There are no user-visible changes.

## How to test 🧪
Create weblish and glish sessions against a host. It should work exactly as before for core regions. In the browser Developer Tools, you should see accesses to the `lish` endpoint and _not_ the `lish_token` endpoint.

### Prerequisites
No special setup is required.

### Reproduction steps
N/A

### Verification steps
See "How to test" above.

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [ ] 🧪 Providing/Improving test coverage
- [x] 🔐 Removing all sensitive information from the code and PR description
- [ ] 🚩 Using a feature flag to protect the release
- [x] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

---
## Commit message and pull request title format standards

> **Note**: Remove this section before opening the pull request
**Make sure your PR title and commit message on squash and merge are as shown below**

`<commit type>: [JIRA-ticket-number] - <description>`

**Commit Types:**
- `feat`: New feature for the user (not a part of the code, or ci, ...).
- `fix`: Bugfix for the user (not a fix to build something, ...).
- `change`: Modifying an existing visual UI instance. Such as a component or a feature.
- `refactor`: Restructuring existing code without changing its external behavior or visual UI. Typically to improve readability, maintainability, and performance.
- `test`: New tests or changes to existing tests. Does not change the production code.
- `upcoming`: A new feature that is in progress, not visible to users yet, and usually behind a feature flag.

**Example:** `feat: [M3-1234] - Allow user to view their login history`

---

### CODE REVIEW

File: packages/manager/src/features/Lish/Lish.tsx [hunk 1]
```diff
@@ -110,7 +113,7 @@ const Lish = () => {
     );
   }
 
-  return linode && token ? (
+  return linode && weblish_url && glish_url && monitor_url && ws_protocols !== undefined ? (
```
Comment (hana-linode) [inline packages/manager/src/features/Lish/Lish.tsx:116]: We should be able to just do 
```suggestion
  return linode && data !== undefined ? (
```

and remove lines 48-51

File: packages/manager/src/features/Lish/Lish.tsx [hunk 2]
```diff
@@ -123,11 +126,11 @@ const Lish = () => {
       <TabLinkList tabs={tabs} />
       <TabPanels>
         <SafeTabPanel data-qa-tab="Weblish" index={0}>
-          <Weblish linode={linode} refreshToken={refreshToken} token={token} />
+          <Weblish linode={linode} refreshToken={refreshToken} weblish_url={weblish_url} ws_protocols={ws_protocols} />
```
Comment (hana-linode) [inline packages/manager/src/features/Lish/Lish.tsx:129]: Then just pass in data like
```suggestion
          <Weblish linode={linode} refreshToken={refreshToken} {...data} />
```

File: packages/manager/src/features/Lish/Lish.tsx [hunk 3]
```diff
@@ -123,11 +126,11 @@ const Lish = () => {
...
         </SafeTabPanel>
         {!isBareMetal && (
           <SafeTabPanel data-qa-tab="Glish" index={1}>
-            <Glish linode={linode} refreshToken={refreshToken} token={token} />
+            <Glish linode={linode} refreshToken={refreshToken} glish_url={glish_url} monitor_url={monitor_url} ws_protocols={ws_protocols} />
```
Comment (hana-linode) [inline packages/manager/src/features/Lish/Lish.tsx:133]:
```suggestion
            <Glish linode={linode} refreshToken={refreshToken} {...data} />
```

File: packages/manager/src/queries/linodes/linodes.ts [hunk 4]
```diff
@@ -130,10 +130,12 @@ export const useLinodeKernelQuery = (kernel: string) => {
   );
 };
 
-export const useLinodeLishTokenQuery = (id: number) => {
-  return useQuery<{ lish_token: string }, APIError[]>(
-    [queryKey, 'linode', id, 'lish-token'],
-    () => getLinodeLishToken(id),
+export const useLinodeLishQuery = (id: number) => {
+  return useQuery<{ weblish_url: string, glish_url: string,
+                    monitor_url: string, ws_protocols: string[] },
```
Comment (hana-linode) [inline packages/manager/src/queries/linodes/linodes.ts:135]: Refer to new interface here

**Review by hana-linode — COMMENTED** [review]
Use new `lish` API ✅ 
Core regions load lish/glish the same as prod ✅ 
New endpoint returns WebSocket protocols for distributed regions ✅ 
WebSocket protocols are empty for core regions ✅ 

Is it expected that the lish console is blank for distributed regions? Granted it doesn't work for me in prod either (just displays a loading spinner)

| Distributed region  | Core region   |
| ------- | ------- |
| Screenshot 2024-07-09 at 12 37 19 PM | Screenshot 2024-07-09 at 12 38 13 PM |

File: packages/api-v4/src/linodes/linodes.ts [hunk 5]
```diff
@@ -29,16 +29,17 @@ export const getLinode = (linodeId: number) =>
...
- * Generates a token which can be used to authenticate with LISH.
+ * Generates urls and websockets protocols to authenticate with LISH.
  *
  * @param linodeId { number } The id of the Linode.
  */
-export const getLinodeLishToken = (linodeId: number) =>
-  Request<{ lish_token: string }>(
+export const getLinodeLish = (linodeId: number) =>
+  Request<{ weblish_url: string, glish_url: string, monitor_url: string,
+            ws_protocols: string[] }>(
```
[the remark this answers is not in the document]
Comment (plisiecki1) [reply packages/api-v4/src/linodes/linodes.ts]: I was unable to get eslint working, but I did address the specific issue above.

**Comment by plisiecki1** [comment]
It is expected that "distributed" will just spin because the backend configuration is not yet 100% complete.

**Comment by hana-linode** [comment]
@plisiecki1 It seems like lish requests are being spammed for the distributed regions on this branch

https://github.com/linode/manager/assets/115299789/008298b6-ebb6-4a0b-8f81-fc18e9760546

**Review by bnussman-akamai — APPROVED** [review]

**Review by mjac0bs — APPROVED** [review]
Observed the `lish` network request as expected in dev tools for weblish and glish, with empty `protocols` for core regions, and session protocol for a distributed region.

Aside from Hana's last observation, feedback addressed + Hana's changes looked good to me.

**Comment by plisiecki1** [comment]
@mjac0bs @bnussman-akamai @hana-linode - I have attempted to address the "spamming" of requests by improving the error/retry logic so that we can retry only several times within any short period of time. Is this too much to take in this change? We could go with the previously reviewed version and then take on the other improvements later on if that seems like a better approach.

**Comment by hana-linode** [comment]
@plisiecki1 These are a lot of new changes, let's go back to the previous version and merge that, then open a follow up PR for improvements

**Comment by plisiecki1** [comment]
I just force pushed to revert the additional improvements, which we can work on in a separate follow-up PR. The code is now exactly the same commit which was previously reviewed.

**Review by hana-linode — APPROVED** [review]

**Pull request merged**